### PR TITLE
feat: track direct API calls in usage analytics (fixes #28059)

### LINKED ISSUE #28059: feat: Requests to the chat completions API from Open Web UI should be included in usage evaluation on an equal footing with normal chat messages.
### Check Existing Issues

- [x] I have searched all existing **open AND closed** issues and discussions and found none comparable to my request.

### Verify Feature Scope

- [x] I believe this feature request is appropriately scoped for the Issues section as described above.

### Problem Description

Currently, I am using the chat completions endpoint provided by Open Web UI to connect language models to third-party AI systems. On Open Web UI's usage analytics page, requests made via the API of individual users are not counted. It would be desirable to include API-based model usage in this analysis.

### Proposed Solution

Requests to the API provided by openwebui should also be considered in the evaluation, just like chat requests via the web interface.

### Alternatives Considered

_No response_

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by SushantGautam**
# Pull Request Checklist

- [x] **Linked Issue/Discussion:** Closes #28059
- [x] **Target branch:** Targets `dev`
- [x] **Description:** Provided below
- [x] **Testing:** 28 automated tests pass; manual curl test verified
- [x] **No Unchecked AI Code:** Changes reviewed and manually tested
- [x] **Self-Review:** Performed
- [x] **Git Hygiene:** 3 atomic commits, rebased on `dev`
- [x] **Title Prefix:** `feat:`

# Changelog Entry

### Description

Direct calls to `/api/chat/completions` / `/api/v1/chat/completions` without a UI chat session are not recorded in analytics because both streaming and non-streaming response handlers wrap tracking inside `if event_emitter:` — pure API calls have `event_emitter = None` (no WebSocket). This PR fixes the silent drop and surfaces API usage in the admin dashboard.

Related: #23245, #23926, #24217, #23341

### Added

- `source` column on `chat_message` (`NULL` = chat message, `'api'` = direct API call). Alembic migration drops FK on `chat_id` (nullable for API rows) and adds column.
- Five admin-only analytics endpoints: `/api/v1/analytics/api/{models,users,summary,daily,tokens}`
- Source dropdown (Chat / API) in Admin → Analytics dashboard; API view hides "Chats" column, relabels "Requests"
- 28 ORM-level tests (upsert, source isolation, token aggregation, date filtering, edge cases)

### Fixed

- `non_streaming_chat_response_handler`: tracking moved outside `if event_emitter:` guard
- `streaming_chat_response_handler` / `stream_wrapper`: usage accumulated from SSE chunks, written after stream exhaustion

### Changed

- All existing chat analytics methods filter `source IS NULL` — existing numbers unaffected

### Contributor License Agreement

- [x] By submitting this pull request, I confirm that I have read and fully agree to the [Contributor License Agreement (CLA)](https://github.com/open-webui/open-webui/blob/main/CONTRIBUTOR_LICENSE_AGREEMENT), and I am providing my contributions under its terms.

### CODE REVIEW

File: backend/open_webui/utils/middleware.py [hunk 1]
```diff
@@ -3692,6 +3693,23 @@ async def non_streaming_chat_response_handler(response, ctx):
                                 **({'usage': usage} if usage else {}),
                             },
                         )
+                    elif not getattr(request.state, 'internal', False) and usage:
+                        # Direct API call (no UI chat session) — record for analytics.
+                        model_id = form_data.get('model') or metadata.get('model_id') or (metadata.get('model') or {}).get('id')
+                        if model_id and user:
```
[the remark this answers is not in the document]
Comment (SushantGautam) [reply backend/open_webui/utils/middleware.py:3700]: Fixed — added `form_data = ctx['form_data']` at the top of `non_streaming_chat_response_handler` alongside the other ctx extractions. Same root cause applied to the suppressed comment on line 3763.

File: backend/open_webui/utils/middleware.py [hunk 2]
```diff
@@ -5640,12 +5696,34 @@ def wrap_item(item):
                 if data:
                     if has_api_outlet_filters:
                         update_assistant_message_from_stream(assistant_message, data)
+                    # Capture usage from the final chunk (stream_options.include_usage)
+                    if isinstance(data, dict) and data.get('usage'):
+                        _stream_usage = data['usage']
                     yield data
```
[the remark this answers is not in the document]
Comment (SushantGautam) [reply backend/open_webui/utils/middleware.py]: Fixed — the loop now handles both dict chunks and raw SSE strings/bytes. For string items we split on newlines, check for `data: ` prefix (skipping `[DONE]`), parse the JSON, and capture `.usage` if present.

File: src/lib/apis/analytics/index.ts [hunk 3]
```diff
@@ -240,6 +240,80 @@ export const getTokenUsage = async (
...
+		.then(async (res) => {
+			if (!res.ok) throw await res.json();
+			return res.json();
+		})
+		.catch((err) => {
+			console.error(err);
+			return null;
+		});
+	return res;
+};
```
[the remark this answers is not in the document]
Comment (SushantGautam) [reply src/lib/apis/analytics/index.ts:272]: Fixed — `_apiFetch` now follows the same pattern as other helpers: captures `err.detail` into an `error` variable and throws after the fetch, so callers receive the error rather than a silent null.

File: backend/open_webui/migrations/versions/a1f3e2d9c8b7_add_source_to_chat_message.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+    conn = op.get_bind()
+    inspector = sa.inspect(conn)
+    dialect = conn.dialect.name
+    columns = {c['name'] for c in inspector.get_columns('chat_message')}
+
+    # Add source column if not present
+    if 'source' not in columns:
+        op.add_column('chat_message', sa.Column('source', sa.Text(), nullable=True, index=True))
+
+    # Make chat_id nullable and drop FK constraint.
```
[the remark this answers is not in the document]
Comment (SushantGautam) [reply backend/open_webui/migrations/versions/a1f3e2d9c8b7_add_source_to_chat_message.py]: Fixed — replaced `index=True` on `op.add_column` with an explicit `op.create_index('ix_chat_message_source', 'chat_message', ['source'])` call (guarded by an existence check against `inspector.get_indexes`).

File: backend/open_webui/models/chat_messages.py [hunk 5]
```diff
@@ -129,8 +129,13 @@ class ChatMessage(Base):
 
     # Identity
     id = Column(Text, primary_key=True)
-    chat_id = Column(Text, ForeignKey('chat.id', ondelete='CASCADE'), nullable=False, index=True)
+    # chat_id is nullable: NULL means the row was created by a direct API call
+    # (no UI chat session). FK intentionally omitted so API-origin rows
+    # are not tied to a Chat row.
+    chat_id = Column(Text, nullable=True, index=True)
```
[the remark this answers is not in the document]
Comment (SushantGautam) [reply backend/open_webui/models/chat_messages.py]: Good catch — agreed the FK drop was too aggressive. Updated the migration to only call `op.alter_column` (making `chat_id` nullable) without touching the FK. Non-NULL rows keep ON DELETE CASCADE; NULL API rows are exempt from the FK check. The ORM column definition omits the FK declaration (no `relationship()` to Chat is needed) to avoid metadata resolution issues, but the constraint lives in the DB via migration.

File: src/lib/components/admin/Analytics/Dashboard.svelte [hunk 6]
```diff
@@ -119,15 +127,38 @@
...
+					: getSummary(localStorage.token, start, end, selectedGroupId),
+				isApi
+					? getApiModelAnalytics(localStorage.token, start, end, selectedGroupId)
+					: getModelAnalytics(localStorage.token, start, end, selectedGroupId),
+				isApi
+					? getApiUserAnalytics(localStorage.token, start, end, 50, selectedGroupId)
+					: getUserAnalytics(localStorage.token, start, end, 50, selectedGroupId),
+				isApi
+					? getApiDailyStats(localStorage.token, start, end, selectedGroupId)
+					: getDailyStats(localStorage.token, start, end, granularity, selectedGroupId),
```
[the remark this answers is not in the document]
Comment (SushantGautam) [reply src/lib/components/admin/Analytics/Dashboard.svelte:143]: Fixed — `getApiDailyStats` now accepts a `granularity` parameter and forwards it as a query param. The backend `/api/daily` endpoint now checks `granularity == 'hourly'` and calls the new `get_api_hourly_counts_by_model` method (mirroring the chat daily/hourly split), so the 24h period shows hourly buckets consistently for both Chat and API sources.
